server/src/main/java/io/druid/server/http/NamespacesResource.java

```diff
@@ -0,0 +1,242 @@
+                        }
+                        return jsonMapper.readValue(
+                            data, new TypeReference<Map<String, Object>>()
+                            {
+                            }
+                        );
+                      }
+                      catch (Exception e) {
+                        log.warn(e, "Error on path [%s]", path);
+                        return null;
+                      }
+                    }
+                  }
+              ), new Predicate<Map<String, Object>>()
+              {
+                @Override
+                public boolean apply(@Nullable Map<String, Object> input)
+                {
+                  return input != null;
+                }
+              }
+          )
+      );
+    }
+    catch (Exception e) {
+      log.error(e, "Error getting list of namespaces");
+      return Response.serverError().entity(cleanError(e)).build();
+    }
+    return builder.build();
+  }
+
+  @POST
+  @Produces({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE})
+  @Consumes({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE})
+  public Response newNamespace(
+      InputStream inputStream,
+      @Context final HttpServletRequest req // used only to get request content-type
+  )
+  {
+    final String reqContentType = req.getContentType();
+    final boolean isSmile = SmileMediaTypes.APPLICATION_JACKSON_SMILE.equals(reqContentType);
+    final ObjectMapper mapper = isSmile ? smileMapper : jsonMapper;
+
+    Response.ResponseBuilder builder = Response.status(Response.Status.ACCEPTED);
+
+
+    final Map<String, Object> map;
+    try {
+      map = mapper.readValue(
+          inputStream, new TypeReference<Map<String, Object>>()
+          {
+          }
+      );
+    }
+    catch (IOException e) {
+      log.error(e, "Error parsing new namespace");
+      return Response.status(Response.Status.BAD_REQUEST).entity(cleanError(e)).build();
+    }
+
+    try {
+      if (!map.containsKey("namespace")) {
+        return Response
+            .status(Response.Status.BAD_REQUEST)
+            .entity(cleanError(new IAE("`namespace` field not found")))
+            .build();
+      }
+      final Object nsObj = map.get("namespace");
+      if (!(nsObj instanceof String)) {
+        return Response.status(Response.Status.BAD_REQUEST).entity(
+            cleanError(
+                new IAE(
+                    String.format(
+                        "`namespace` field is not String, is [%s]", nsObj.getClass().getCanonicalName()
+                    )
+                )
+            )
+        ).build();
+      }
+      final String namespace = (String) nsObj;
+      final String path = ZKPaths.makePath(zkPathsConfig.getNamespacePath(), namespace);
+      if(curator.checkExists().forPath(path) == null) {
+        curator
+            .inTransaction()
+            .create()
+            .withMode(CreateMode.PERSISTENT)
```

Comment:
Using persistent ZK storage now changes the ZK dependence that we have.  Before this change, it was legal to drop your quorum, fire up a new one from scratch and bring up a Druid cluster around it without any change in functionality.  Now we need to actually have backups of the ZK state and stuff as well.  Please store this in the metadata store instead of persistent nodes on ZK.